src/polkit/polkitunixprocess.c

```diff
@@ -1131,6 +1362,445 @@ get_start_time_for_pid (pid_t    pid,
+  dev_t minor;
+  struct stat st;
+  struct statfs stfs;
+  gchar *tty_name;
+  gchar *filename;
+  gchar *contents;
+  size_t length;
+  gchar **tokens;
+  guint num_tokens;
+  gchar *p;
+  gchar *endp;
+
+  tty_nr = 0;
+  tty_fd = -1;
+  pid = 0;
+  tty_name = NULL;
+  contents = NULL;
+  tokens = NULL;
+
+  g_return_val_if_fail (pidfd >= 0, 0);
+
+  pid = polkit_unix_process_get_pid_from_pidfd (pidfd, error);
+  if (pid <= 0)
+    return 0;
+
+  filename = g_strdup_printf ("/proc/%d/stat", pid);
+  if (!filename)
+    {
+      g_set_error (error,
+                   POLKIT_ERROR,
+                   POLKIT_ERROR_FAILED,
+                   "Error allocating memory for procfs path");
+      goto out;
+    }
+
+  if (!g_file_get_contents (filename, &contents, &length, error))
+    goto out;
+
+  /* parent's pid is the token at index 1 after the '(process name)' entry - since only this
+   * field can contain the ')' character, search backwards for this to avoid malicious
+   * processes trying to fool us
+   */
+  p = strrchr (contents, ')');
+  if (p == NULL)
+    {
+      g_set_error (error,
+                   POLKIT_ERROR,
+                   POLKIT_ERROR_FAILED,
+                   "Error parsing file %s",
+                   filename);
+      goto out;
+    }
+  p += 2; /* skip ') ' */
+  if (p - contents >= (int) length)
+    {
+      g_set_error (error,
+                   POLKIT_ERROR,
+                   POLKIT_ERROR_FAILED,
+                   "Error parsing file %s",
+                   filename);
+      goto out;
+    }
+
+  tokens = g_strsplit (p, " ", 0);
+
+  num_tokens = g_strv_length (tokens);
+
+  if (num_tokens < 5)
+    {
+      g_set_error (error,
+                   POLKIT_ERROR,
+                   POLKIT_ERROR_FAILED,
+                   "Error parsing file %s",
+                   filename);
+      goto out;
+    }
+
+  /*
+   * The string can have negative representation, but it's actually an unsigned int,
+   * cast it to convert it
+   */
+  tty_nr = (guint) strtoll (tokens[4], &endp, 10);
+  if (endp == tokens[4])
+    {
+      g_set_error (error,
+                   POLKIT_ERROR,
+                   POLKIT_ERROR_FAILED,
+                   "Error parsing file %s",
+                   filename);
+      goto out;
+    }
+
+  if (tty_nr == 0)
+    {
+      g_set_error (error,
+                   POLKIT_ERROR,
+                   POLKIT_ERROR_FAILED,
+                   "No TTY associated with process %d",
+                   pid);
+      goto out;
+    }
+
+  /* Now do some sanity checks to ensure we are really looking at a TTY. */
+
+  /*
+   * As per the manpage, the minor device number is contained in the
+   * combination of bits 31 to 20 and 7 to 0, the major device number is in
+   * bits 15 to 8.
+   */
+  major = (tty_nr & 0xff00) >> 8;
+  minor = ((tty_nr & 0xfff00000) >> 12) | (tty_nr & 0xff);
+
+  tty_name = g_strdup_printf ("/dev/pts/%lu", minor);
+
+  tty_fd = open (tty_name, O_RDONLY|O_CLOEXEC);
```

Comment:
oh, don't open random ttys like this. ttys are magic shit. at the very least you want O_NOCTTY, but note that opening ttys can be blocking too, you shouldn't just do that. use O_PATH, if you just care about th einode.

but frankly, this all is unnecessary. you can trust the kernel on only allowing ttys to become cttys, this validation is nonsensical.

also note that ptys are getting reused, hence reopening them iva /dev/pts is a mistake.